### PR TITLE
Enable File.Copy to EXFAT volumes when not running as root

### LINKED ISSUE #45730: FileInfo.CopyTo UnauthorizedAccessException Error on Copying File to exFAT Drive in Linux
Hello,

Im trying to copy some files within an external exfat drive. As we all know, exfat in linux ubuntu belongs to root and has 777 permission so the copy should just work without using `sudo` or any elevated user. However when i use `FileInfo.Copy` it gives me:
```
System.UnauthorizedAccessException: Access to the path is denied.\n ---> System.IO.IOException: Operation not permitted\n   --- End of inner exception stack trace ---\n   at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirectory, Func`2 errorRewriter)\n   at Interop.CheckIo(Int64 result, String path, Boolean isDirectory, Func`2 errorRewriter)\n   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)\n   at System.IO.FileInfo.CopyTo(String destFileName, Boolean overwrite)\n   at System.Threading.Tasks.Task`1.InnerInvoke()\n   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)\n--- End of stack trace from previous location ---\n   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)\n--- End of stack trace from previous location ---\n   at home_server.Services.SystemFile.CopyDirectoryAsync(String src, String target) in /home/x/deploy/home-server/home-server/Services/SystemFile.cs:line 130\n   at home_server.Services.CameraRecordingArchivalService.CreateAsync(Camera camera, ISystemFile markerFile, ISystemFile recordingDir, ISystemFile archivalDir) in /home/x/deploy/home-server/home-server/Services/CameraRecordingArchivalService.cs:line 121\n   at home_server.Jobs.CameraRecordingArchivalJob.DoWork() in /home/x/deploy/home-server/home-server/Jobs/CameraRecordingArchivalJob.cs:line 49"
```
 
Which shouldnt happen as i can just copy, paste, create, write to that drive anywhere without `sudo` despite of the owner of everything in that drive is `root`.

To proof this here is the function i use instead of `FileInfo.Copy` and this works just fine:
```
async Task CopyFileAsync(string src, string target)
        {
            using (Stream srcFs = File.Open(src, FileMode.Open))
            using (Stream targetFs = File.Open(target, FileMode.Create))
            {
                await srcFs.CopyToAsync(targetFs);
            }
        }
```

I believe this is a bug and not an expected behavior. Im sorry if this issue has been posted before. Thanks in advance.

---
```
dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.100
 Commit:    5044b93829

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  20.04
 OS Platform: Linux
 RID:         ubuntu.20.04-arm64
 Base Path:   /usr/share/dotnet/sdk/5.0.100/

Host (useful for support):
  Version: 5.0.0
  Commit:  cf258a14b7

.NET SDKs installed:
  3.1.301 [/usr/share/dotnet/sdk]
  5.0.100 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.5 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.5 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
```
Target Framework: `net5.0`

Ivan

### PR DESCRIPTION
**Pull request by danmoseley**
Fixes https://github.com/dotnet/runtime/issues/45730

In File.Copy on Linux, after successfully copying the file we copy over the access and modified time, and copy the access permissions. See #16366 for why this was done: it's because on Windows, CopyFileEx preserves them, and some programs took a dependency on this e.g., for up to dateness checks, and arguably it's always been documented that File.Copy copies 'attributes'.

The problem on exfat volumes is that this file system does not store POSIX ownership; when viewed from Linux, all files appear to be owned by root. When we copy to such a volume, the copy succeeds if we have write permissions, but subsequently setting the timestamps and access permissions fails because (unless running as root) we need to own the file, and only root owns it. This manifests as `System.IO.IOException: Operation not permitted.` even though the file is successfully copied. 

The workaround if you want to copy to an exfat volume on Linux is to either implement File.Copy yourself, or run as root.

The fix is to ignore EPERM when setting the time stamps and permissions. That gives a reasonable outcome, the file is copied in the best way that we can achieve given the filesystem limitations.

Note that "touch" does succeed against files on exfat if you do not supply a time -- the man page for utimensat notes that to set current time only requires write access, to set a specific time requires ownership rights or sufficient privileges (eg., root) -- and on exfat, only root ever has ownership rights. Conceivably we could at least set the current time, therefore, but it's not clear that's an improvement.

I don't think there's a good way to write a test but I tested manually on an exfat volume.

### CODE REVIEW

**Comment by danmosemsft** [comment]
Test failure is https://github.com/dotnet/runtime/issues/44657

**Comment by danmosemsft** [comment]
This also impacts FileInfo.CopyTo

**Comment by danmosemsft** [comment]
@Jozkee any chance you could take a quick look?

File: src/libraries/Native/Unix/System.Native/pal_io.c [hunk 1]
```diff
@@ -1165,14 +1165,17 @@ int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destinationFd)
...
+    // copy metadata. The best thing we can do is skip copying the metadata.
+    if (ret != 0 && errno != EPERM)
     {
         return -1;
     }
-
     // Then copy permissions.
     while ((ret = fchmod(outFd, sourceStat.st_mode & (S_IRWXU | S_IRWXG | S_IRWXO))) < 0 && errno == EINTR);
-    if (ret != 0)
+    if (ret != 0 && errno != EPERM) // See EPERM comment above
```
Comment (Jozkee) [inline src/libraries/Native/Unix/System.Native/pal_io.c:1178]: Does fchmod also fails with EPERM when running the same exFAT scenario?

**Review by jozkee — APPROVED** [review]
Sounds reasonable to me. Thanks!

> I don't think there's a good way to write a test but I tested manually on an exfat volume.

Not required for this PR but if you think is valuable to have the manual test you could consider adding it to https://github.com/dotnet/runtime/blob/master/src/libraries/System.IO.FileSystem/tests/ManualTests/ManualTests.cs

File: src/libraries/Native/Unix/System.Native/pal_io.c (hunk 1 above)
Comment (danmosemsft) [reply src/libraries/Native/Unix/System.Native/pal_io.c:1178]: Yes I verified it does - that's in the man pages too

**Comment by Jozkee** [comment]
@danmosemsft forgot to ask, is there a problem if you run the same scenario on Windows? I wonder if this might cause a platform inconsistency.

**Comment by danmosemsft** [comment]
@Jozkee that should not be a problem, as on Windows we don't copy the attributes -- Windows itself does, inside CopyFileEx. That's what our Unix implementation is trying to mimic.

**Comment by danmosemsft** [comment]
added manual test

**Pull request merged**